microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceASTValidator.java

```diff
@@ -204,6 +222,53 @@ private void validateAsynchronousAnnotation(MethodDeclaration node, MarkerAnnota
 		}
 	}
 
+	/**
+	 * Checks if the given method declaration has a retry annotation, and if so,
+	 * provides diagnostics for the delay and maxDuration value(s)
+	 *
+	 * @param annotation The @Retry annotation
+	 * @throws JavaModelException
+	 */
+	private void validateRetryAnnotation(NormalAnnotation annotation) throws JavaModelException {
+		Expression delayExpr = getAnnotationMemberValueExpression(annotation, DELAY_RETRY_ANNOTATION_MEMBER);
+		Expression maxDurationExpr = getAnnotationMemberValueExpression(annotation,
+				MAX_DURATION_RETRY_ANNOTATION_MEMBER);
+		if (delayExpr != null && maxDurationExpr != null) {
+
+			Expression delayUnitExpr = getAnnotationMemberValueExpression(annotation,
+					DELAY_UNIT_RETRY_ANNOTATION_MEMBER);
+			Expression durationUnitExpr = getAnnotationMemberValueExpression(annotation,
+					DURATION_UNIT_RETRY_ANNOTATION_MEMBER);
+			Expression jitterExpr = getAnnotationMemberValueExpression(annotation, JITTER_RETRY_ANNOTATION_MEMBER);
+			Expression jitterDelayUnitExpr = getAnnotationMemberValueExpression(annotation,
+					JITTER_DELAY_UNIT_RETRY_ANNOTATION_MEMBER);
+
+			int delayNum = (int) delayExpr.resolveConstantExpressionValue();
+			int maxDurationNum = (int) maxDurationExpr.resolveConstantExpressionValue();
+			int jitterNum = jitterExpr != null ? (int) jitterExpr.resolveConstantExpressionValue() : 0;
+
+			String delayUnit = delayUnitExpr != null ? delayUnitExpr.toString().split("\\.")[1] : null;
+			String maxDurationUnit = durationUnitExpr != null ? durationUnitExpr.toString().split("\\.")[1] : null;
+			String jitterDelayUnit = jitterDelayUnitExpr != null ? jitterDelayUnitExpr.toString().split("\\.")[1]
+					: null;
+
+			Duration delayValue = delayUnitExpr != null ? Duration.of(delayNum, ChronoUnit.valueOf(delayUnit))
+					: Duration.of(delayNum, ChronoUnit.MILLIS);
+			Duration maxDurationValue = durationUnitExpr != null
+					? Duration.of(maxDurationNum, ChronoUnit.valueOf(maxDurationUnit))
+					: Duration.of(maxDurationNum, ChronoUnit.MILLIS);
+			Duration jitterValue = jitterDelayUnitExpr != null
+					? Duration.of(jitterNum, ChronoUnit.valueOf(jitterDelayUnit))
+					: Duration.of(jitterNum, ChronoUnit.MILLIS);
+			Duration maxDelayValue = delayValue.plus(jitterValue);

```

Comment:
Ok, jitter it is. It isn't just added to the delay, but also subtracted. It adds some random delay to the minium value. Have a look at https://github.com/eclipse/microprofile-fault-tolerance/blob/800363fe002701bb2e31ba413f21e5a36af56b41/api/src/main/java/org/eclipse/microprofile/faulttolerance/Retry.java#L113 .

The reason I hesitated is because if you take a delay of 3, with a jitter of 2, and maxDuration of 4, you'll have delays between [1,5] . Some of the delays will fall below 4, but others above. Should we all this an error ? Maybe a warning ?

Reply:
Currently the error message reports "The `delay` member value must be less than the `maxDuration` member value.", but to illustrate this, I can mention "the sum of `delay` and `jitter`...". 

According to https://github.com/eclipse/microprofile-fault-tolerance/blob/800363fe002701bb2e31ba413f21e5a36af56b41/api/src/main/java/org/eclipse/microprofile/faulttolerance/Retry.java#L113, the negative effective delays have a lower bound of 0, which means the maximum effective delay should be the only one we need to validate.